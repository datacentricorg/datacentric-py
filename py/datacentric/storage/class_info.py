# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, List
import typing_inspect


class ClassInfo:
    """Contains reflection based helper static methods.
    """
    __is_initialized: bool = False
    __data_types_map: Dict[str, type] = dict()

    @staticmethod
    def get_type(name: str) -> type:
        """Returns data derived type given its name."""
        if not ClassInfo.__is_initialized:
            from datacentric.storage.data import Data
            children = ClassInfo.__get_runtime_imported_data(Data, [])
            for child in children:
                if child not in ClassInfo.__data_types_map:
                    ClassInfo.__data_types_map[child.__name__] = child
            ClassInfo.__is_initialized = True

        if name not in ClassInfo.__data_types_map:
            raise Exception(f'Class {name} is not found in ClassInfo data types map')

        return ClassInfo.__data_types_map[name]

    @staticmethod
    def get_ultimate_base(type_: type) -> type:
        """
        Returns the ultimate base class of the inheritance chain which
        determines the collection name.

        This is the class derived directly from Data, Record, or RootRecord.
        """
        from datacentric.storage.data import Data
        from datacentric.storage.record import Record
        from datacentric.storage.root_record import RootRecord
        root_types = [Data, Record, RootRecord]

        if type_.mro()[0] in root_types:
            raise Exception(f'Cannot get root type from root type.')
        type_mro = type_.mro()
        for root_type in root_types:
            if root_type in type_mro:
                index = type_mro.index(root_type)
                return type_mro[index - 1]
        raise Exception(f'Type is not derived from Data, Record, or RootRecord.')

    @staticmethod
    def __get_runtime_imported_data(type_: type, children: List[type]) -> List[type]:
        """For the given type recursively adds its children."""
        current_children = type_.__subclasses__()
        for t in current_children:
            ClassInfo.__get_runtime_imported_data(t, children)
        children.extend(current_children)
        return children
