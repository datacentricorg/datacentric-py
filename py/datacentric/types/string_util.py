# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import re
from typing import Pattern


class StringUtil():
    """
    Static helper class for str.
    """

    __first_cap_re: Pattern = re.compile('(.)([A-Z][a-z]+)')
    __all_cap_re: Pattern = re.compile('([a-z0-9])([A-Z])')

    @staticmethod
    def to_pascal_case(name: str) -> str:
        """
        Converts strings to PascalCase also removing underscores.
        """
        result: str = ''.join(x for x in name.title() if not x == '_')
        return result

    @staticmethod
    def to_snake_case(name: str) -> str:
        """
        Converts PascalCase strings to snake_case.
        """
        s1: str = StringUtil.__first_cap_re.sub(r'\1_\2', name)
        result: str = StringUtil.__all_cap_re.sub(r'\1_\2', s1).lower()
        return result
