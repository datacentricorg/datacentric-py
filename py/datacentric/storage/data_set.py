# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import attr
from bson import ObjectId
from typing import List, Union
from datacentric.storage.record import Record
from datacentric.storage.data_set_key import DataSetKey


@attr.s(slots=True, auto_attribs=True)
class DataSet(Record):
    """
    Dataset is a concept similar to a folder, applied to data in any
    data source including relational or document databases, OData
    endpoints, etc.

    Datasets can be stored in other datasets. The dataset where dataset
    record is stored is called parent dataset.

    Dataset has an Imports array which provides the list of TemporalIds of
    datasets where records are looked up if they are not found in the
    current dataset. The specific lookup rules are specific to the data
    source type and described in detail in the data source documentation.

    Some data source types do not support Imports. If such data
    source is used with a dataset where Imports array is not empty,
    an error will be raised.

    The root dataset uses TemporalId.Empty and does not have versions
    or its own DataSet record. It is always last in the dataset
    lookup sequence. The root dataset cannot have Imports.
    """

    data_set_name: str = attr.ib(default=None, kw_only=True)
    """Unique dataset name."""

    non_temporal: bool = attr.ib(default=None, kw_only=True, metadata={'optional': True})
    """
    Flag indicating that the dataset is non-temporal even if the
    data source supports temporal data.

    For the data stored in datasets where NonTemporal == false, a
    temporal data source keeps permanent history of changes to each
    record within the dataset, and provides the ability to access
    the record as of the specified TemporalId, where TemporalId serves
    as a timeline (records created later have greater TemporalId than
    records created earlier).

    For the data stored in datasets where NonTemporal == true, the
    data source keeps only the latest version of the record. All
    child datasets of a non-temporal dataset must also be non-temporal.

    In a non-temporal data source, this flag is ignored as all
    datasets in such data source are non-temporal.
    """

    imports: List[ObjectId] = attr.ib(default=None, kw_only=True, repr=False, metadata={'optional': True})
    """
    List of datasets where records are looked up if they are
    not found in the current dataset.

    The specific lookup rules are specific to the data source
    type and described in detail in the data source documentation.

    The parent dataset is not included in the list of Imports by
    default and must be included in the list of Imports explicitly.
    """

    def to_key(self) -> str:
        """Get DataSet key."""
        return 'DataSet=' + self.data_set_name

    @classmethod
    def create_key(cls, *, data_set_name: str) -> Union[str, DataSetKey]:
        """Create DataSet key."""
        return 'DataSet=' + data_set_name
