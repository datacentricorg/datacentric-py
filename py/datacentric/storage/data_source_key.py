# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import attr
from typing import Union, Optional, List, Any
from abc import ABC, abstractmethod
from datacentric.storage.key import Key


class DataSourceKey(Key):
    """
    Data source is a logical concept similar to database
    that can be implemented for a document DB, relational DB,
    key-value store, or filesystem.

    Data source API provides the ability to:

    (a) store and query datasets;
    (b) store records in a specific dataset; and
    (c) query record across a group of datasets.

    This record is stored in root dataset.
    """
    pass
