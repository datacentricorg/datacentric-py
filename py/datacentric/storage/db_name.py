# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional
from abc import ABC
from datacentric.storage.typed_key import TypedKey
from datacentric.storage.root_record import RootRecord
from datacentric.storage.instance_type import EnvType


class DbNameKey(TypedKey['DbName']):
    """
    This class enforces strict naming conventions
    for database naming. While format of the resulting database
    name is specific to data store type, it always consists
    of three tokens: InstanceType, InstanceName, and EnvName.
    The meaning of InstanceName and EnvName tokens depends on
    the value of InstanceType enumeration.

    This record is stored in root dataset.
    """

    __slots__ = ('instance_type', 'instance_name', 'env_name')

    instance_type: EnvType
    instance_name: Optional[str]
    env_name: Optional[str]

    def __init__(self):
        super().__init__()

        self.instance_type = None
        """
        Instance type enumeration.

        Some API functions are restricted based on the instance type.
        """

        self.instance_name = None
        """
        The meaning of instance name depends on the instance type.

        * For PROD, UAT, and DEV instance types, instance name
        identifies the endpoint.

        * For USER instance type, instance name is user alias.

        * For TEST instance type, instance name is the name of
        the unit test class (test fixture).
        """

        self.env_name = None
        """
        The meaning of environment name depends on the instance type.

        * For PROD, UAT, DEV, and USER instance types, it is the
        name of the user environment selected in the client.

        * For TEST instance type, it is the test method name.
        """


class DbName(RootRecord[DbNameKey], ABC):
    """
    This class enforces strict naming conventions
    for database naming. While format of the resulting database
    name is specific to data store type, it always consists
    of three tokens: InstanceType, InstanceName, and EnvName.
    The meaning of InstanceName and EnvName tokens depends on
    the value of InstanceType enumeration.

    This record is stored in root dataset.
    """

    __slots__ = ('instance_type', 'instance_name', 'env_name')

    instance_type: Optional[EnvType]
    instance_name: Optional[str]
    env_name: Optional[str]

    def __init__(self):
        super().__init__()

        self.instance_type = None
        """
        Instance type enumeration.

        Some API functions are restricted based on the instance type.
        """

        self.instance_name = None
        """
        The meaning of instance name depends on the instance type.

        * For PROD, UAT, and DEV instance types, instance name
        identifies the endpoint.

        * For USER instance type, instance name is user alias.

        * For TEST instance type, instance name is the name of
        the unit test class (test fixture).
        """

        self.env_name = None
        """
        The meaning of environment name depends on the instance type.

        * For PROD, UAT, DEV, and USER instance types, it is the
        name of the user environment selected in the client.

        * For TEST instance type, it is the test method name.
        """
