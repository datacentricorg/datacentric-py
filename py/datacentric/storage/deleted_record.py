# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import attr
from datacentric.storage.record import Record


@attr.s(slots=True)
class DeletedRecord(Record):
    """
    When returned by the data source, this record has the same
    effect as if no record was found. It is used to indicate
    a deleted record when audit log must be preserved.
    """

    _key: str = attr.ib(default=None, kw_only=True)
    """Attribute to hold key."""

    @property
    def key(self) -> str:
        """
        String key consists of semicolon delimited primary key elements:

        KeyElement1;KeyElement2

        To avoid serialization format uncertainty, key elements
        can have any atomic type except Double.
        """

        return self._key

    @key.setter
    def key(self, value: str) -> None:
        self._key = value

    def to_key(self) -> str:
        """Deleted record key."""
        return self._key
