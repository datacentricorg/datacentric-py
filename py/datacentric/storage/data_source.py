# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod
from typing import Optional, TypeVar, Iterable, List
from bson import ObjectId

from datacentric.storage.data_set_flags import DataSetFlags
from datacentric.storage.typed_key import TypedKey
from datacentric.storage.record import Record
from datacentric.storage.typed_record import TypedRecord
from datacentric.storage.data_set import DataSet
from datacentric.storage.env_type import EnvType

TRecord = TypeVar('TRecord', bound=Record)


class DataSourceKey(TypedKey['DataSource']):
    """
    Data source is a logical concept similar to database
    that can be implemented for a document DB, relational DB,
    key-value store, or filesystem.

    Data source API provides the ability to:

    (a) store and query datasets;
    (b) store records in a specific dataset; and
    (c) query record across a group of datasets.

    This record is stored in root dataset.
    """

    __slots__ = ('data_source_name',)

    data_source_name: Optional[str]

    def __init__(self):
        super().__init__()

        self.data_source_name = None
        """Unique data source name."""


class DataSource(TypedRecord[DataSourceKey], ABC):
    """
    Data source is a logical concept similar to database
    that can be implemented for a document DB, relational DB,
    key-value store, or filesystem.

    Data source API provides the ability to:

    (a) store and query datasets;
    (b) store records in a specific dataset; and
    (c) query record across a group of datasets.

    This record is stored in root dataset.
    """

    __slots__ = ('data_source_name', 'env_type', 'env_group', 'env_name', 'non_temporal', 'read_only')

    _empty_id = ObjectId('000000000000000000000000')
    common_id: str = 'Common'

    data_source_name: Optional[str]
    env_type: Optional[EnvType]
    env_group: Optional[str]
    env_name: Optional[str]
    non_temporal: bool
    read_only: bool

    def __init__(self):
        super().__init__()

        self.data_source_name = None
        """Unique data source name."""

        self.env_type = None
        """
        Environment type enumeration.

        Some API functions are restricted based on the environment type.
        """

        self.env_group = None
        """
        The meaning of environment group depends on the environment type.

        * For PROD, UAT, and DEV environment types, environment group
        identifies the endpoint.

        * For USER environment type, environment group is user alias.

        * For TEST environment type, environment group is TestCaseName,
        the unique identifier of test case record.
        """

        self.env_name = None
        """
        The meaning of environment name depends on the environment type.

        * For PROD, UAT, DEV, and USER environment types, it is the
        name of the user environment selected in the client.

        * For TEST environment type, it is the test method name.
        """

        self.non_temporal = None
        """
        Flag indicating that the data source is non-temporal.

        For the data stored in data sources where NonTemporal == false,
        the data source keeps permanent history of changes to each
        record (except where dataset or record are marked as NonTemporal),
        and provides the ability to access the record as of the specified
        TemporalId, where TemporalId serves as a timeline (records created
        later have greater TemporalId than records created earlier).

        For the data stored in data source where NonTemporal == true,
        the data source keeps only the latest version of the record. All
        datasets created by a NonTemporal data source must also be non-temporal.

        In a non-temporal data source, this flag is ignored as all
        datasets in such data source are non-temporal.
        """

        self.read_only = None
        """
        Use this flag to mark data source as readonly.

        Data source may also be readonly because CutoffTime is set.
        """

    @abstractmethod
    def create_ordered_object_id(self) -> ObjectId:
        """The returned ObjectIds have the following order guarantees:

        * For this data source instance, to arbitrary resolution; and
        * Across all processes and machines, to one second resolution

        One second resolution means that two ObjectIds created within
        the same second by different instances of the data source
        class may not be ordered chronologically unless they are at
        least one second apart."""
        pass

    @abstractmethod
    def load_or_null(self, record_type: type, id_: ObjectId) -> Optional[TRecord]:
        """Load record by its ObjectId.

        Return None if there is no record for the specified ObjectId;
        however an exception will be thrown if the record exists but
        is not derived from type_.
        """
        pass

    @abstractmethod
    def load_or_null_by_key(self, key_: TypedKey, load_from: ObjectId) -> Optional[TRecord]:
        """Load record by string key from the specified dataset or
        its list of imports. The lookup occurs first in descending
        order of dataset ObjectIds, and then in the descending
        order of record ObjectIds within the first dataset that
        has at least one record. Both dataset and record ObjectIds
        are ordered chronologically to one second resolution,
        and are unique within the database server or cluster.

        The root dataset has empty ObjectId value that is less
        than any other ObjectId value. Accordingly, the root
        dataset is the last one in the lookup order of datasets.

        The first record in this lookup order is returned, or null
        if no records are found or if DeletedRecord is the first
        record.

        Return None if there is no record for the specified ObjectId;
        however an exception will be thrown if the record exists but
        is not derived from TRecord.
        """
        pass

    @abstractmethod
    def get_query(self, record_type: type, load_from: ObjectId):
        """Get query for the specified type.

        After applying query parameters, the lookup occurs first in
        descending order of dataset ObjectIds, and then in the descending
        order of record ObjectIds within the first dataset that
        has at least one record. Both dataset and record ObjectIds
        are ordered chronologically to one second resolution,
        and are unique within the database server or cluster.

        The root dataset has empty ObjectId value that is less
        than any other ObjectId value. Accordingly, the root
        dataset is the last one in the lookup order of datasets.
        """
        pass

    @abstractmethod
    def save_many(self, record_type: type, records: Iterable[TRecord], save_to: ObjectId) -> None:
        """Save multiple records to the specified dataset. After the method exits,
        for each record the property record.data_set will be set to the value of
        the save_to parameter.

        All save methods ignore the value of record.data_set before the
        save method is called. When dataset is not specified explicitly,
        the value of dataset from the context, not from the record, is used.
        The reason for this behavior is that the record may be stored from
        a different dataset than the one where it is used.

        This method guarantees that ObjectIds of the saved records will be in
        strictly increasing order.
        """
        pass

    @abstractmethod
    def delete(self, key: TypedKey[TRecord], delete_in: ObjectId) -> None:
        """Write a DeletedRecord in delete_in dataset for the specified key
        instead of actually deleting the record. This ensures that
        a record in another dataset does not become visible during
        lookup in a sequence of datasets.

        To avoid an additional roundtrip to the data store, the delete
        marker is written even when the record does not exist.
        """
        pass

    @abstractmethod
    def delete_db(self) -> None:
        """Permanently deletes (drops) the database with all records
        in it without the possibility to recover them later.

        This method should only be used to free storage. For
        all other purposes, methods that preserve history should
        be used.

        ATTENTION - THIS METHOD WILL DELETE ALL DATA WITHOUT
        THE POSSIBILITY OF RECOVERY. USE WITH CAUTION.
        """
        pass

    @abstractmethod
    def get_data_set_or_none(self, data_set_name: str, load_from: ObjectId) -> Optional[ObjectId]:
        """Get ObjectId of the dataset with the specified name.
        Returns null if not found.
        """
        pass

    @abstractmethod
    def save_data_set(self, data_set: DataSet, save_to: ObjectId) -> None:
        """Save new version of the dataset and update in-memory cache to the saved dataset."""
        pass

    # From extensions:
    def load(self, record_type: type, id_: ObjectId) -> TRecord:
        """Load record by its ObjectId.

        Error message if there is no record for the specified ObjectId,
        or if the record exists but is not derived from record_type.
        """
        raise NotImplemented

    def load_by_key(self, key_: TypedKey, load_from: ObjectId) -> TRecord:
        """Load record from context.data_source, overriding the dataset
        specified in the context with the value specified as the
        second parameter. The lookup occurs in the specified dataset
        and its imports, expanded to arbitrary depth with repetitions
        and cyclic references removed.

        IMPORTANT - this overload of the method loads from load_from
        dataset, not from context.data_set.

        Error message if the record is not found or is a DeletedRecord.
        """
        result = self.load_or_null_by_key(key_, load_from)
        if result is None:
            raise Exception(f'Record with key {key_} is not found in dataset with TemporalId={load_from}.')
        return result

    def save_one(self, record_type: type, record: TRecord, save_to: ObjectId):
        """Save one record to the specified dataset. After the method exits,
        record.data_set will be set to the value of the data_set parameter.

        All save methods ignore the value of record.data_set before the
        Save method is called. When dataset is not specified explicitly,
        the value of dataset from the context, not from the record, is used.
        The reason for this behavior is that the record may be stored from
        a different dataset than the one where it is used.

        This method guarantees that ObjectIds of the saved records will be in
        strictly increasing order.
        """
        self.save_many(record_type, [record], save_to)

    def get_common(self) -> ObjectId:
        """Return ObjectId of the latest Common dataset.
        Common dataset is always stored in root dataset.
        """
        return self.get_data_set(DataSource.common_id, DataSource._empty_id)

    def get_data_set(self, data_set_name: str, load_from: ObjectId) -> ObjectId:
        """Get ObjectId of the dataset with the specified name.
        Error message if not found.
        """
        result = self.get_data_set_or_none(data_set_name, load_from)
        if result is None:
            raise Exception(f'Dataset {data_set_name} is not found in data store {self.data_source_name}.')
        return result

    def create_common(self, flags: DataSetFlags = None) -> ObjectId:
        """Create Common dataset with the specified flags.

        The flags may be used, among other things, to specify
        that the created dataset will be NonTemporal even if the
        data source is itself temporal. This setting is typically
        used to prevent the accumulation of data where history is
        not needed.

        By convention, the Common dataset contains reference and
        configuration data and is included as import in all other
        datasets.

        The Common dataset is always stored in root dataset.

        This method updates in-memory dataset cache to include
        the created dataset.
        """
        if flags is None:
            flags = DataSetFlags.Default
        return self.create_data_set(DataSource.common_id, DataSource._empty_id, flags=flags)

    def create_data_set(self, data_set_name: str, parent_data_set: ObjectId, imports: List[ObjectId] = None,
                        flags: DataSetFlags = None) -> ObjectId:
        """Create dataset with the specified data_set_name, parent_data_set, imports,
        and flags.

        This method updates in-memory dataset cache to include
        the created dataset.
        """
        if flags is None:
            flags = DataSetFlags.Default
        if imports is None:
            imports = [parent_data_set]

        result = DataSet()
        result.data_set_name = data_set_name
        if imports is not None:
            result.imports = [x for x in imports]

        if (self.non_temporal is not None and self.non_temporal) or \
                flags & DataSetFlags.NonTemporal == DataSetFlags.NonTemporal:
            result.non_temporal = True

        self.save_data_set(result, parent_data_set)

        return result.id_
