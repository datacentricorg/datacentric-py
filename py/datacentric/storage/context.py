# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Optional
import inspect
from bson.objectid import ObjectId

# To prevent linter error on type hint in quotes
if False:
    from datacentric.logging.log import Log
    from datacentric.storage.data_source import DataSource


class Context:
    """
    Context provides:

    * Default data source
    * Default dataset of the default data source
    * Logging
    * Progress reporting
    * Filesystem access (if available)
    """

    __log: Optional['Log']
    data_source: Optional['DataSource']
    data_set: Optional[ObjectId]

    def __init__(self):

        self.__log = None
        """Logging interface."""

        self.data_source = None
        """Default data source of the context."""

        self.data_set = None
        """Default dataset of the context."""

    # --- PROPERTIES

    @property
    def log(self) -> 'Log':
        """Logging interface."""

        # Define log
        from datacentric.logging.log import Log

        if not self.__log:
            raise Exception('fLog property is not set in {GetType().Name}.')
        return self.__log

    @log.setter
    def log(self, value: 'Log'):
        """Logging interface."""
        self.__log = value
        self.__log.init(self)

    # --- METHODS

    def configure(self, module) -> None:
        """
        Invokes static method configure(context) method with self as argument
        for every class that is loaded by the argument and marked with
        ConfigurableAttribue class attribute.

        The method configure(context) may be used to configure:

        * Reference data, and
        * In case of test mocks, test data

        The order in which configure(context) method is invoked when
        multiple classes marked by ConfigurableAttribue are present
        is undefined. The implementation of configure(context) should
        not rely on any existing data, and should not invoke other
        Configure(context) method of other classes.

        The attribute ConfigurableAttribue is not inherited. To invoke
        Configure(context) method for multiple classes within the same
        inheritance chain, specify ConfigurableAttribue for each
        class that provides configure(context) method.
        """

        for name, obj in inspect.getmembers(module, lambda x: inspect.isclass(x)):
            print(name)
            print(inspect.getmembers(obj))
