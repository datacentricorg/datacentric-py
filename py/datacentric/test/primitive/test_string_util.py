# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
from datacentric.storage.context import Context
from datacentric.primitive.string_util import StringUtil


class TestStringUtil(unittest.TestCase):
    """Unit tests for StringUtil."""

    def test_smoke(self):
        """Smoke test"""
        context: Context = self.create_method_context()

        # Does not remove space TODO - check that Humanizer does in C#
        context.log.verify(StringUtil.to_pascal_case('abc def'))
        context.log.verify(StringUtil.to_pascal_case('abc_def'))

        # Does not remove space TODO - check that Humanizer does in C#
        context.log.verify(StringUtil.to_snake_case('Abc Def'))
        context.log.verify(StringUtil.to_snake_case('AbcDef'))


if __name__ == "__main__":
    unittest.main()
