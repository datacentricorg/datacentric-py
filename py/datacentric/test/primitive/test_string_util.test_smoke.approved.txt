Verify: Abc Def
Verify: AbcDef
Verify: abc _def
Verify: abc_def
Verify: Test completed successfully.
